graph/patterns/change-status.md

```diff
@@ -0,0 +1,86 @@
+# Change Status & Action Request
+
+Microsoft Graph API Design Pattern
+
+_The change status pattern is an effective alternative to model the ability to trigger side effects while keeping an agnostic, data driven approach_
+
+## Problem
+
+In some cases, exposed resources can trigger side effects that modify the system behavior. For instance, a [`riskyUser`](https://docs.microsoft.com/en-us/graph/api/resources/riskyuser?view=graph-rest-1.0) can be [`confirmed`](https://docs.microsoft.com/en-us/graph/api/riskyuser-confirmcompromised?view=graph-rest-1.0&tabs=http) or [`dismissed`](https://docs.microsoft.com/en-us/graph/api/riskyuser-confirmcompromised?view=graph-rest-1.0&tabs=http).
```

Comment:
I would be all for it, but I have not found any - and I am afraid that this is the way things have been modeled in Graph since forever. I will take another look, but I am not confident I can find something

Reply:
Actually thinking about this again, I think it's good to use this example to show how it should have been done.  [Aside: Potentially the API can be updated without a breaking change, and the actions can be deprecated.]

If we are using this as an example, then I would recommend showing more of the request and response in your examples below.

```json
PATCH /identityProtection/riskyUsers/{id}
{
   "desiredRiskState": "confirmedCompromised"
}

200 OK
{
   "id": "{id}",
   "riskState": "confirmedCompromised",
   "desiredRiskState": "confirmedCompromised",
 }

This would be the "synchronous" operation.
Alternatively, we could just say that riskState itself is updateable, but only for setting dismissed or confirmedCompromised values.  All other values are system set values and hence are read-only - setting those would be a 400 error. 

The challenge here is that the existing actions are also bulk operations (changing the state of multiple users in one request).  We should put this to one side for the time-being IMHO.

We might actually want to have 2 examples (using riskyUsers) - one where the operation is synchronous, and one where the operation is long-running.  While we can point to the LRO docs, I think it's good to see how these might be combined.

The other example to point to (which is much more complex) is the model that PIM and Access Reviews use - where there is a request object for creating the actual object.  https://docs.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagementv3-overview?view=graph-rest-1.0 and https://docs.microsoft.com/en-us/graph/api/resources/accesspackageassignmentrequest?view=graph-rest-1.0

Reply:
> Potentially the API can be updated without a breaking change, and the actions can be deprecated

If this gets passed, I would love to work with some "early adopter" workload to try this out and see if that fits the use case better!

> The challenge here is that the existing actions are also bulk operations (changing the state of multiple users in one request). We should put this to one side for the time-being IMHO.

I would be open to work on this use case as well, but to be honest I would need some evidence that the system is either significantly slower if doing X requests instead of a single request with an array of X user ids or some good scenario/customer complain.

To expand a little bit my line of thinking, if you have an operation on a single object, you don’t need to define another version just to operate on a collection of those objects.
 

I give the developer `X`, and they can do `map collection Y over X`. That is not something I should be providing.

Having a separate definition for processing a group of things implies that there is something special about processing a group instead of a single item. If that is the case (for example, a more efficient batch implementation.) I might (not) be open to consider it. Let me know if that makes sense or not.